### PR TITLE
Allow to override sharded jedis pooled object factory

### PR DESCRIPTION
**Pull request by bokshitsky**
Hi,

I want to be able to override default ShardedJedisFactory from ShardedJedisPool.

main reason: current validateObject() method uses PING-PONG mechanism that looks quite expensive when setTestOnReturn(true)/setTestOnBorrow(true) are set and a lot of shards are used.

I want to try something like:
```
@Override
public boolean validateObject(PooledObject<ShardedJedis> pooledShardedJedis) {
  try {
    ShardedJedis jedis = pooledShardedJedis.getObject();
    for (Jedis shard : jedis.getAllShards()) {
      if (shard.isBroken() || !shard.isConnected()) {
        return false;
      }
    }
    return true;
  } catch (RuntimeException ex) {
    return false;
  }
}
```
for our projects at https://github.com/hhru

But I am afraid what merging same code in current implementation is quite dangerous due to possible backward compatibility broke.

So I add required constructor to ShardedJedisPool and make ShardedJedisFactory public to allow in-place override. (which is tested by new unit-test)

Also fixed some code-smells (catch generic Exception instead of just Runtime, make slowlog test more stable by truncating initial state)

Feel free to ask any questions.
Excuse me for my english - I am not native speaker :)

### CODE REVIEW

File: src/test/java/redis/clients/jedis/tests/commands/SlowlogCommandsTest.java [hunk 1]
```diff
@@ -36,6 +36,7 @@ public void tearDown() throws Exception {
 
   @Test
   public void slowlog() {
+    jedis.slowlogReset();
```
Comment (sazzad16) [inline src/test/java/redis/clients/jedis/tests/commands/SlowlogCommandsTest.java:39]: If you think this is necessary, do this in a separate PR.

File: src/test/java/redis/clients/jedis/tests/ShardedJedisPoolOverrideObjectFactoryTest.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import redis.clients.jedis.*;
+import redis.clients.jedis.util.Hashing;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.junit.Assert.assertEquals;
+
+public class ShardedJedisPoolOverrideObjectFactoryTest {
```
Comment (sazzad16) [inline src/test/java/redis/clients/jedis/tests/ShardedJedisPoolOverrideObjectFactoryTest.java:16]: I don't see any necessity of creating a separate test class. Add the test in the existing one.

File: src/main/java/redis/clients/jedis/ShardedJedisPool.java [hunk 3]
```diff
@@ -83,12 +88,12 @@ public void destroyObject(PooledObject<ShardedJedis> pooledShardedJedis) throws
             if (!jedis.isBroken()) {
               jedis.quit();
             }
-          } catch (Exception e) {
+          } catch (RuntimeException e) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/ShardedJedisPool.java:91]: Exception was caught so that every node gets a chance to be destroyed even though there is a bug. This is an operation that MUST be done (as much as possible).

File: src/main/java/redis/clients/jedis/ShardedJedisPool.java [hunk 4]
```diff
@@ -83,12 +88,12 @@ public void destroyObject(PooledObject<ShardedJedis> pooledShardedJedis) throws
...
             logger.warn("Error while QUIT", e);
           }
           try {
             jedis.disconnect();
-          } catch (Exception e) {
+          } catch (RuntimeException e) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/ShardedJedisPool.java:96]: -do-

File: src/main/java/redis/clients/jedis/ShardedJedisPool.java [hunk 5]
```diff
@@ -105,7 +110,7 @@ public boolean validateObject(PooledObject<ShardedJedis> pooledShardedJedis) {
           }
         }
         return true;
-      } catch (Exception ex) {
+      } catch (RuntimeException ex) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/ShardedJedisPool.java:113]: -do-

**Review by sazzad16 — CHANGES_REQUESTED** [review]

File: src/main/java/redis/clients/jedis/ShardedJedisPool.java (hunk 3 above)
Comment (bokshitsky) [reply src/main/java/redis/clients/jedis/ShardedJedisPool.java:91]: Right now checked `Exception` is not thrown from the code inside - so my proposal is backward-compatible 
(I change Exception -> RuntimeException, not Throwable -> RutimeExceptio)

but catching generic checked `Exception` is quite "dangerous" if internal contract will change and something like `Interrupted Exception` will be thrown: current code will not handle it properly - but someone who make changes won't get compilation error and may miss this (as a result - thread won't be interrupted)

Comment (bokshitsky) [reply src/main/java/redis/clients/jedis/ShardedJedisPool.java:91]: > Exception was caught so that every node gets a chance to be destroyed even though there is a bug. This is an operation that MUST be done (as much as possible).

Probably I don't understood something what you mean, 
may you please explain some details how Exception->RuntimeException breaks current execution logic?

File: src/test/java/redis/clients/jedis/tests/ShardedJedisPoolOverrideObjectFactoryTest.java (hunk 2 above)
Comment (bokshitsky) [reply src/test/java/redis/clients/jedis/tests/ShardedJedisPoolOverrideObjectFactoryTest.java:16]: 👍

File: src/test/java/redis/clients/jedis/tests/commands/SlowlogCommandsTest.java (hunk 1 above)
Comment (bokshitsky) [reply src/test/java/redis/clients/jedis/tests/commands/SlowlogCommandsTest.java:39]: 👍

**Comment by bokshitsky** [comment]
> If you think this is necessary, do this in a separate PR.

> I don't see any necessity of creating a separate test class. Add the test in the existing one.

I am quite newbie in contributing - how should I fix these comments?
1 - push some commits to current PR and squash whem in one later?
2 - or just remove/change current commits locally and then force-push them?
 
 I like way 2)

**Comment by sazzad16** [comment]
@bokshitsky 
1. you can push more commits. generally we (maintainers) squash them.
2. force push is okay as well (specially in smaller PRs).

File: src/test/java/redis/clients/jedis/tests/commands/SlowlogCommandsTest.java (hunk 1 above)
Comment (bokshitsky) [reply src/test/java/redis/clients/jedis/tests/commands/SlowlogCommandsTest.java]: done - removed this change from PR

File: src/test/java/redis/clients/jedis/tests/ShardedJedisPoolOverrideObjectFactoryTest.java (hunk 2 above)
Comment (bokshitsky) [reply src/test/java/redis/clients/jedis/tests/ShardedJedisPoolOverrideObjectFactoryTest.java]: done - moved changes to current test class

File: src/main/java/redis/clients/jedis/ShardedJedisPool.java (hunk 3 above)
Comment (sazzad16) [reply src/main/java/redis/clients/jedis/ShardedJedisPool.java:91]: @bokshitsky At lease separate these changes in another PR with explanation.

**Review by sazzad16 — CHANGES_REQUESTED** [review]

**Comment by bokshitsky** [comment]
> At lease separate these changes in another PR with explanation.

done, removed this commit

**Review by sazzad16 — APPROVED** [review]

**Review by yangbodong22011 — APPROVED** [review]

**Comment by bokshitsky** [comment]
Thank you for merging my branch.

could you please tell something about jedis release policy - when the next release is planned?

I ask because if it will be soon I will wait for version in mvncentral.
If the release is not planned soon i will build version myself locally and put it to company local nexus.

**Comment by sazzad16** [comment]
@bokshitsky There is no recent release plan.
